weasyprint/draw/text.py

```diff
@@ -172,15 +173,40 @@ def draw_first_line(stream, textbox, text_overflow, block_ellipsis, matrix):
             string = ''
             stream.set_font_size(font.hash, 1 if font.bitmap else font_size)
         string += '<'
-        for i in range(num_glyphs):
-            glyph_info = glyphs[i]
-            glyph = glyph_info.glyph
+        for i in range(number_of_glyphs):
+            glyph_info = glyphs_info[i]
+            glyph_id = glyph_info.glyph
             width = glyph_info.geometry.width
-            if (glyph == pango.PANGO_GLYPH_EMPTY or
-                    glyph & pango.PANGO_GLYPH_UNKNOWN_FLAG):
+
+            # Display zero-width empty glyph.
+            if glyph_id == pango.PANGO_GLYPH_EMPTY:
                 string += f'>{-width / font_size}<'
                 continue
 
+            # Display .notdef for missing glyphs.
+            if glyph_id & pango.PANGO_GLYPH_UNKNOWN_FLAG:
+                codepoint = glyph_id - pango.PANGO_GLYPH_UNKNOWN_FLAG
+                glyph_id = font.get_unused_glyph_id(codepoint)
+                font.widths[glyph_id] = round(width * 1000 * FROM_UNITS / font_size)
+                if 0 not in font.widths:
+                    # "width" is actually Pango’s get_approximate_char_width. Force
+                    # .notdef’s to use this width, even if it’s not the right, as we
+                    # want to keep Pango’s layout for next glyphs.
+                    font.widths[0] = font.widths[glyph_id]
+
+            # Create mapping between glyphs and Unicode codepoints.
+            if glyph_id not in font.to_unicode:
+                utf8_slice = slice(*sorted(utf8_positions[i:i+2]))
+                text = font.to_unicode[glyph_id] = utf8_text[utf8_slice].decode()
+
+            # Log warning for missing glyphs.
+            if glyph_id & pango.PANGO_GLYPH_UNKNOWN_FLAG:
+                codepoints = ', '.join(f'U+{ord(character):04X}' for character in text)
+                LOGGER.warning(
+                    'Missing glyph when displaying "%s", Unicode codepoint(s) %s',
```

Comment:
Besides moving this logging to precede `glyph_id = font.get_unused_glyph_id(codepoint)`, there’s also some editorial suggestion:

```suggestion
                    '.notdef glyph rendered for Unicode string unsupported by fonts: "%s" (%s)',
```